### ISSUE TITLE
ID from endpoint is getting changed

### ISSUE THREAD

**Issue by ethaneisenhard** [danspratling/gatsby-source-multi-api#2]
I love this source plugin, but i think that the code is changing the ID of the data coming in if you use this line

const nodeId = createNodeId(rest-api-result-${result.id})

which makes it hard to run a PUT HTTP request back to the data endpoint if you want to update it, I think you might want to remove that line or make it optional

I am trying to see if I can use the code you wrote and just remove that part but I am not sure how I can do that in my own project, any help would be greatly appreciated.

**Comment by ethaneisenhard** [comment]
Update: 

I have made a pull request, not sure if it will work, because i dont know how to test it, its mostly pseudo code, but i think you will get the idea, let me know if you have questions

**Comment by danspratling** [comment]
I'm not able to replicate this with my test API. Node ID's shouldn't be able to change because the ID is generated from the result ID (which also shouldn't change).

If you want to test the above you can fork the project and then include it as a package in your project using 

```
// package.json

"dependencies": {
    "gatsby-source-rest-api": "../gatsby-source-rest-api", //reference to your local file
}
```

If you build the package `yarn build` and then restart your test project you should be able to see any changes you make.

It'd also be helpful if you could set up a minimum reproducible test - a Gatsby project including this plugin and only the other things required to cause this bug so i can take a look and see what's causing it.

This is an unfinished plugin so it's definitely possible there are bugs but i don't want to make changes where i can't replicate the result.

**Comment by danspratling** [comment]
#3

**Comment by ethaneisenhard** [comment]
It may because of the DB i am using, I have a repo set up at 

https://github.com/ethaneisenhard/calendar-app branch example 

with the My JSON server DB at
https://github.com/ethaneisenhard/calendarappdb

You can view the endpoint data here
https://my-json-server.typicode.com/ethaneisenhard/calendarappdb/events

so you can see ID = 1

so that will work, but if you change the plugin to the original code 
change gatsby-node, line 22 to 
const nodeId = createNodeId(rest-api-result-${result.id})

You will see in the graphQL that the ID changes from 1 to createNodeId()

**Comment by danspratling** [comment]
Okay so after some discussion via DMs we came to the conclusion that this was 2 issues stemming from 1 use case.

> **Use case**: The intention is that a user can submit data (in this case a calendar post) back to the database and the ID needs to be present so that data can be overwritten. 

Currently, the ID is being morphed into a GUID from something like "1" to "4df5409b-7b76-5189-b3ea-926aab75efb8", but the original ID of the endpoint isn't being preserved for use. This has been fixed in #4 and now both the morphed GUID and original ID are present on the returned data object.

____

A separate use case which was also found was that the data was not able to update in real time, which isn't a fault of the plugin (though not being able to reference the original id would definitely hinder this). This can be resolved by a few different means. The way gatsby sources data is via node.js, which is a server side language and can't handle real-time data management. It should only be used for the initial data load.

To solve this, I would suggest creating a state object with your data in once the page has loaded and then once the user makes an update you can update that state with new data.

This should get you started in the right direction. Fetch can handle getting and updating data so something like this will be a good starting point.

```
import React, { useState } from 'react'

export default ({ data }) => {
  const [
    userData, //stored data for the user
    setUserData //method to update the user data & refresh state
  ] = useState(data.users)

  const deleteUser = id => {
    fetch(`http://example.com/api/users/${id}`, {
        method: 'DELETE', //Can be PUT, POST, DELETE, 
    })
      .then(response => response.json())
      .then(data => console.log(data)) //this probably returns nothing for delete, you'd likely want to do a second fetch to get the new data set. Up to you how you want to approach this with the different methods
      .catch(error => console.log(error))
  }

//will also need a way to handle Add & Update, and also updating the state with the new data once the user data has changed, which i haven't handled here. This was just to give you a push in the right direction if you needed it!

  return (
    <button onClick={() => deleteUser(1)}>Delete user</button>
  )
}

//graphQL query 
```

Alternately, you could use something like RxJS to help you handle this https://rxjs-dev.firebaseapp.com/guide/overview
